cylc/flow/commands.py

```diff
@@ -0,0 +1,434 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Cylc scheduler commands.
+
+These are the scripts which are actioned on the Scheduler instance when you
+call a mutation.
+
+Each is an async generator providing functionalities for:
+
+* Validation:
+  * The generator is executed up to the first "yield" before the command is
+    queued.
+  * Validation and argument parsing can be performed at this stage.
+  * If the generator raises an Exception then the error message will be
+    communicated back to the user and the command will not be queued.
+  * If the execution string is not obvious to a user, catch the exception and
+    re-raise it as an InputError with a more obvious string.
+  * Any other exceptions will be treated as genuine errors.
+* Execution:
+  * The generator is executed up to the second "yield" when the command is run
+    by the Scheduler's main-loop:
+  * The execution may also stop at a return or the end of the generator code.
+  * If the generator raises a CommandFailedError at this stage, the error will
+    be caught and logged.
+  * Any other exceptions will be treated as genuine errors.
+
+In the future we may change this interface to allow generators to "yield" any
+arbitrary number of strings to serve the function of communicating command
+progress back to the user. For example, the generator might yield the messages:
+
+* Command queued.
+* Done 1/3 things
+* Done 2/3 things
+* Done 3/3 things
+* Done
+
+For more info see: https://github.com/cylc/cylc-flow/issues/3329
+
+"""
+
+from contextlib import suppress
+from time import sleep, time
+from typing import (
+    AsyncGenerator,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    TYPE_CHECKING,
+    Union,
+)
+
+from cylc.flow import LOG
+import cylc.flow.command_validation as validate
+from cylc.flow.exceptions import (
+    CommandFailedError,
+    CyclingError,
+    CylcConfigError,
+)
+import cylc.flow.flags
+from cylc.flow.log_level import log_level_to_verbosity
+from cylc.flow.network.schema import WorkflowStopMode
+from cylc.flow.parsec.exceptions import ParsecError
+from cylc.flow.task_id import TaskID
+from cylc.flow.task_state import TASK_STATUSES_ACTIVE, TASK_STATUS_FAILED
+from cylc.flow.workflow_status import RunMode, StopMode
+
+from metomi.isodatetime.parsers import TimePointParser
+
+if TYPE_CHECKING:
+    from cylc.flow.scheduler import Scheduler
+
+    # define a type for command implementations
+    Command = Callable[
+        ...,
+        AsyncGenerator,
+    ]
+
+# a directory of registered commands (populated on module import)
+COMMANDS: 'Dict[str, Command]' = {}
+
+
+def _command(name: str):
+    """Decorator to register a command."""
+    def _command(fcn: 'Command'):
+        nonlocal name
+        COMMANDS[name] = fcn
+        fcn.command_name = name  # type: ignore
+        return fcn
+    return _command
```

Comment:
```suggestion
def _command(name: str = None):
    """Decorator to register a command."""
    def _command(fcn: 'Command'):
        nonlocal name
        if not name:
            name = fcn.__name__
        COMMANDS[name] = fcn
        fcn.command_name = name  # type: ignore
        return fcn
    return _command
```

Reply:
?

Reply:
Crucial line 

```
name = name if name else fcn.__name__
```

So that if you are happy with the function name (and I think it's only set, where we're not) you can do

```python
@_command
async def sensible_name():
```

Reply:
What issue are you trying to resolve?

Reply:
Just thought the decorator would be neater if you didn't have to give it an arg unless you really needed to.

Reply:
Ah, I wanted the mutation:command mapping to be explicit to decouple these concepts so this duplication is purposeful.

Reply:
FYI: This is what I wanted:
 
```python
from cylc.flow.network.schema import Mutations

@command(Mutations.stop)
def stop(schd ...):
```

It's doable, but nasty due to Graphene internals not attaching the field name to the mutation.